lib/datadog/appsec/utils/read_write_lock.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+module Datadog
+  module AppSec
+    module Utils
+      # Simple implentation for a fair read write lock
+      # It allows for as many concurrent readers but only one concurrent writer.
+      # To avoid the writer thread being exhausted, once a writer thread has signaled that it wants to acquire the lock,
+      # no more reader threads can acquire the lock.
+      # The new reader threads will wait for the writer thread to finish.
+      class ReadWriteLock
+        def initialize
+          @reader_mutex    = Mutex.new
+          @reader_q        = ConditionVariable.new
+          @reader_count    = 0
+          @reader_releases = 0
+
+          @writer_mutex    = Mutex.new
+          @writer_q        = ConditionVariable.new
+          @writter         = false
+        end
+
+        def with_rlock
+          rlock
+          yield
+        ensure
+          runlock
+        end
+
+        def with_lock
+          lock
+          yield
+        ensure
+          unlock
+        end
+
+        def rlock
+          @reader_mutex.synchronize do
+            @reader_count += 1
+
+            writter = false
+
+            @writer_mutex.synchronize do
+              writter = @writter
+            end
+
+            @reader_q.wait(@reader_mutex) if writter
+          end
+        end
+
+        def runlock
+          @reader_mutex.synchronize do
+            @reader_releases += 1
+
+            @writer_mutex.synchronize { @writer_q.signal } if @reader_releases == @reader_count
+          end
+        end
+
+        def lock
+          @writer_mutex.synchronize do
+            @writer_q.wait(@writer_mutex) while (@reader_releases != @reader_count) || @writter
+
+            @writter = true
+          end
```

Comment:
> Otherwise readers will keep arriving

Indeed it looks like `@reader_count += 1` could happen while a writer is waiting.

> you probably want the writer to set the flag to true before waiting on the condition variable

- It would mean the write lock is taken before waiting?
- The `(@reader_releases != @reader_count) || @writter` condition would always be `true` (since no `unlock` would reset it to `false` because of `@writer_mutex.synchronize`, which would not make sense anyway since that `unlock` would presumably apply to _another_ write lock)

Instead, symmetrically to how the writer lock is taken _after_ waiting, it seems like the read lock should be taken only _after_ waiting, so when no writer is in progress, so it looks like `@reader_count += 1`  should be inside the `@writer_mutex.synchronize do ... end` block of `rlock`.

Reply:
> For that to happen, you probably want the writer to set the flag to true before waiting on the condition variable.

In that case the writter thread will always wait for the condition variable. 

> Instead, symmetrically to how the writer lock is taken after waiting, it seems like the read lock should be taken only after waiting, so when no writer is in progress, so it looks like @reader_count += 1 should be inside the @writer_mutex.synchronize do ... end block of rlock.

Wouldn't make more sense to increment `reader_count` after the wait outside of the `@write_mutex.syncronize`  than inside of it? 

Something like: 

```ruby
def rlock
  @reader_mutex.synchronize do
    writter = false

    @writer_mutex.synchronize do
      writter = @writter
    end

    @reader_q.wait(@reader_mutex) if writter

    @reader_count += 1
  end
end
```

cc @lloeki @ivoanjo